src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,456 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "disk_noise.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <cstring>
+#include <fstream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "mixer.h"
+#include "setup.h"
+#include "timer.h"
+
+// ******************************************************
+// Simulated disk noise for floppy and hard disk drives
+// ******************************************************
+
+// Load a disk noise sample. Expects a WAV file with 16-bit PCM data at 44100 Hz.
+void DiskNoiseDevice::LoadSample(const std::string& path, std::vector<int16_t>& buffer)
+{
+	constexpr auto sample_ext = ".wav";
+
+	if (path.empty()) {
+		return;
+	}
+	// Start with the name as-is and then try from resources
+	const auto candidate_paths = {std_fs::path(path),
+	                              std_fs::path(path + sample_ext),
+	                              get_resource_path(DiskNoiseDir, path),
+	                              get_resource_path(DiskNoiseDir,
+	                                                path + sample_ext)};
+
+	for (const auto& path : candidate_paths) {
+		if (std_fs::exists(path) &&
+		    (std_fs::is_regular_file(path) || std_fs::is_symlink(path))) {
+			std::ifstream file(path, std::ios::binary);
+			if (!file.is_open()) {
+				LOG_WARNING("DiskNoiseDevice: Failed to open %s",
+				            path.c_str());
+				return;
+			}
+
+			// Skip 44 bytes of the WAV header
+			char header[44];
+			file.read(header, 44);
+
+			std::vector<int16_t> temp;
+			int16_t sample;
+			while (file.read(reinterpret_cast<char*>(&sample),
+			                 sizeof(sample))) {
+				temp.push_back(sample);
+			}
+			LOG_DEBUG("DiskNoiseDevice: Loaded %zu samples from %s",
+			          temp.size(),
+			          path.c_str());
+			buffer = std::move(temp);
+			return;
+		}
+	}
+}
+
+void DiskNoiseDevice::LoadSeekSamples(const std::vector<std::string>& paths)
+{
+	for (const auto& path : paths) {
+		if (path.empty()) {
+			seek_samples_.emplace_back(); // Placeholder for missing
+			                              // sample
+			continue;
+		}
+
+		std::vector<int16_t> sample;
+		LoadSample(path, sample);
+		if (!sample.empty()) {
+			seek_samples_.push_back(std::move(sample));
+		} else {
+			seek_samples_.emplace_back();
+		}
+	}
+
+	// Assign weights with descending priority (most weight to first samples)
+	const size_t max_weight = 10;
+	seek_sample_weights_.resize(seek_samples_.size());
+	for (size_t i = 0; i < seek_samples_.size(); ++i) {
+		seek_sample_weights_[i] = static_cast<size_t>(max_weight - i);
+	}
+}
+
+size_t DiskNoiseDevice::ChooseWeightedSeekIndex() const
+{
+	size_t total_weight = 0;
```

Comment:
Using `size_t` for "weight" is a weird choice? You should only use that when interfacing with stdlib functions that expect `size_t`, but otherwise `int` all the way.